src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/Reliability/SqlConfigurableRetryLogicManager.cs

```diff
@@ -15,18 +16,27 @@ internal sealed class SqlConfigurableRetryLogicManager
     {
         private const string TypeName = nameof(SqlConfigurableRetryLogicManager);
 
+#if NET6_0_OR_GREATER
+        [UnconditionalSuppressMessage("Trimming", "IL2026:Members annotated with 'RequiresUnreferencedCodeAttribute' require dynamic access otherwise can break functionality when trimming application code",
+            Justification = "Usages of this property annotated with RequiresUnreferencedCode")]
+#endif
         private static readonly Lazy<SqlConfigurableRetryLogicLoader> s_loader =
-            new Lazy<SqlConfigurableRetryLogicLoader>(() =>
-            {
-                ISqlConfigurableRetryConnectionSection cnnConfig = null;
-                ISqlConfigurableRetryCommandSection cmdConfig = null;
+            new Lazy<SqlConfigurableRetryLogicLoader>(InitializeSqlConfigurableRetryLogicProvider);
+
+#if NET6_0_OR_GREATER
+        [RequiresUnreferencedCode("Calls Microsoft.Data.SqlClient.SqlConfigurableRetryLogicLoader.SqlConfigurableRetryLogicLoader(ISqlConfigurableRetryConnectionSection, ISqlConfigurableRetryCommandSection, String, String)")]
+#endif
+        private static SqlConfigurableRetryLogicLoader InitializeSqlConfigurableRetryLogicProvider()
+        {
+            ISqlConfigurableRetryConnectionSection cnnConfig = null;
+            ISqlConfigurableRetryCommandSection cmdConfig = null;
 
-                // Fetch the section attributes values from the configuration section of the app config file.
-                cnnConfig = AppConfigManager.FetchConfigurationSection<SqlConfigurableRetryConnectionSection>(SqlConfigurableRetryConnectionSection.Name);
-                cmdConfig = AppConfigManager.FetchConfigurationSection<SqlConfigurableRetryCommandSection>(SqlConfigurableRetryCommandSection.Name);
+            // Fetch the section attributes values from the configuration section of the app config file.
+            cnnConfig = AppConfigManager.FetchConfigurationSection<SqlConfigurableRetryConnectionSection>(SqlConfigurableRetryConnectionSection.Name);
+            cmdConfig = AppConfigManager.FetchConfigurationSection<SqlConfigurableRetryCommandSection>(SqlConfigurableRetryCommandSection.Name);
```

Comment:
@Wraith2 starting from this two lines we apply `RequiresUnreferencedCode` across codebase. this bubble up to override of `DbDataReader` like `ExecuteScalar`/`ExecuteDbDataReader`, etc. In .NET DbDataReader does not marked with RUC so it is prohibited to mark these methods with RUC in derived classes. 

I'm not sure that App.config code path is actually excercised on .NET variant, but if so, I really would like to have different configuration for that. if no configuration in app.config then only one [instance of class](https://github.com/dotnet/SqlClient/blob/272219b0428e9dce078cc63c21cf108a94ed9050/src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/Reliability/SqlConfigurableRetryLogicLoader.cs#L134) is available here, so I really would like to streamline instantiation of that class without using all these "reflection" magic somehow.

Reply:
Assuming it's uncommon to use the config file to provide a custom class, this whole thing might be a candidate for a [feature switch](https://github.com/dotnet/designs/blob/main/accepted/2020/feature-switch.md).

Basically disable this capability by default for trimmed apps. If the user enables it explicitly via a property in the csproj file, they would get a warning and would have to deal with it on their own.

But this requires somebody with the knowledge of how this is used - the feature switches are a reasonable solution only for rarely used capabilities. They also degrade the overall experience - instead of providing compile time warnings about trim incompatibilities, the failure will happen only at runtime, so they should be used only when other solutions are not viable.